packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -311,75 +472,132 @@ fn batches_from_shared_store(
     })
 }
 
-fn mem_batch_to_js<'m>(
-    mv8: &'m MiniV8,
+fn mem_batch_to_js<'s>(
+    scope: &mut v8::HandleScope<'s>,
     batch_id: &str,
-    mem: mv8::Object<'m>,
+    mem: v8::Local<'s, v8::Object>,
     persisted: Metaversion,
-) -> Result<mv8::Value<'m>> {
-    let batch = mv8.create_object();
-    let batch_id = mv8.create_string(batch_id);
-    batch.set("id", mv8::Value::String(batch_id))?;
-    batch.set("mem", mem)?;
-    batch.set("mem_version", persisted.memory())?;
-    batch.set("batch_version", persisted.batch())?;
-    Ok(mv8::Value::Object(batch))
+) -> JSRunnerResult<v8::Local<'s, v8::Value>> {
+    let batch = v8::Object::new(scope);
+    let batch_id = new_js_string(scope, batch_id)?;
+
+    let id_field = new_js_string(scope, "id")?;
+    let mem_field = new_js_string(scope, "mem")?;
+    let mem_version_field = new_js_string(scope, "mem_version")?;
+    let batch_version_field = new_js_string(scope, "batch_version")?;
+
+    batch
+        .set(scope, id_field.into(), batch_id.into())
+        .ok_or_else(|| JSRunnerError::V8("Could not set id field on batch".to_string()))?;
+    batch
+        .set(scope, mem_field.into(), mem.into())
+        .ok_or_else(|| JSRunnerError::V8("Could not set mem field on batch".to_string()))?;
+    let js_memory = v8::Number::new(scope, persisted.memory() as f64);
+    batch
+        .set(scope, mem_version_field.into(), js_memory.into())
+        .ok_or_else(|| JSRunnerError::V8("Could not set mem_version field on batch".to_string()))?;
+    let js_batch = v8::Number::new(scope, persisted.batch() as f64);
+    batch
+        .set(scope, batch_version_field.into(), js_batch.into())
+        .ok_or_else(|| {
+            JSRunnerError::V8("Could not set batch_version field on batch".to_string())
+        })?;
+
+    Ok(batch.into())
 }
 
-fn batch_to_js<'m>(
-    mv8: &'m MiniV8,
+fn batch_to_js<'s>(
+    scope: &mut v8::HandleScope<'s>,
     mem: &Memory,
     persisted: Metaversion,
-) -> Result<mv8::Value<'m>> {
-    // TODO: Is `mem.data.len()` different from `mem.size`? (like Vec capacity vs len?)
-    let arraybuffer = mv8.create_arraybuffer(mem.data.as_ptr(), mem.size);
+) -> JSRunnerResult<v8::Local<'s, v8::Value>> {
+    // The memory is owned by the shared memory
+    // we don't want JS or Rust to try to de-allocate it
+    unsafe extern "C" fn no_op(_: *mut std::ffi::c_void, _: usize, _: *mut std::ffi::c_void) {}
+
+    tracing::debug!("Calling batch_to_js");
+    // https://github.com/denoland/rusty_v8/pull/926
```

Comment:
Could you add a `// SAFETY: ` comment, please?

Reply:
I can add some of the safety requirements but I don't know enough about JS/v8/shared memory to make a full list.
For example the pointer has to point to valid memory. But there is definitely some ownership rules that I don't know. Like when is it ok to modify that memory.

Reply:
I think it's fine to just add a section to the actual call and why you think it's safe to do so.